[TF FE] Support Div operation for TensorFlow

sami0i opened the pull request:
### Details:
 - *Support Div operation for TF*

### Tickets:
 - Closes https://github.com/openvinotoolkit/openvino/issues/21464

popovaan commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,48 @@
+        return tf_net, None
+
+    test_data_basic = [
+        dict(input_shape=[10, 20], input_type=np.float32),
```
Can you please add scalar test case?
dict(input_shape=[], input_type=np.float32),

sami0i replied:
Yes, certainly. I will add this scalar test case.

popovaan commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,48 @@
+        inputs_data = {}
+        # generate x and y
+        inputs_data['x'] = np.random.randint(-10, 10, x_shape).astype(self.input_type)
+        inputs_data['y'] = np.random.randint(1, 10, y_shape).astype(self.input_type)
```
Are there any limitation that "y" should be positive? Can we change to `np.random.randint(-10, 10, y_shape)`

sami0i replied:
When the denominator is zero, the output will be inf. Both functions correctly return inf. However, the function that compares them returns false, which causes the test to fail. The `allclose` function in the `common_utils.py` file returns false for two equal values that include inf. Therefore, I excluded zero from the denominator values for the test. Of course, I will include negative values as well.

popovaan replied:
Okay, then let's create positive array and multiply on random sign to have both positive and negative non-zero values.
`np.random.randint(1, 10, y_shape)*np.random.choice([-1,1], y_shape)`

popovaan commented on src/frontends/tensorflow_common/src/op/div.cpp:
```diff
@@ -0,0 +1,29 @@
+namespace frontend {
+namespace tensorflow {
+namespace op {
+OutputVector translate_div_op(const NodeContext& node) {
```
We have a template translator for simple binary ops: https://github.com/openvinotoolkit/openvino/blob/6060683f4c1ac320f1c01fc5c8fc8b8f53f93f02/src/frontends/tensorflow_common/src/op/binary_op.cpp#L63

Let's reuse it for this operation.

sami0i replied:
Could you please show me an example of using this template, if available?

popovaan replied:
You can use `Add` or other operations from this list as example. It should be enough to add a line with `Div` in the list:
https://github.com/openvinotoolkit/openvino/blob/139242d47d6ec8bdb42a05954f51381968435e53/src/frontends/tensorflow_common/src/op/binary_op.cpp#L63

Here is implementation of `translate_binary_op` template: 
https://github.com/openvinotoolkit/openvino/blob/139242d47d6ec8bdb42a05954f51381968435e53/src/frontends/tensorflow_common/src/op/binary_op.cpp#L39

As `Div` translator has no `Div`-specific code I recommend to reuse the template.

sami0i commented:
I have modified `div.cpp `based on the `translate_binary_op` template. However, integrating this structure into the `op_table` requires the use of certain classes defined in the core. Thus, I changed it to `{"Div",CreatorFunction(translate_binary_op(opset8::Divide)}` in the `op_table`.
Currently, there is code for `translate_floor_div_op` in `op_table.cpp`. Can I remove `div.cpp` and implement the div operation in the same manner as `translate_floor_div_op`?

rkazants commented on src/frontends/tensorflow_common/src/op/div.cpp:
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "common_op_table.hpp"
+#include "openvino/op/divide.hpp"
+
+using namespace std;
+using namespace ov::op;
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+namespace op {
+OutputVector translate_binary_op(const NodeContext& node,
+                                 const std::function<shared_ptr<v1::Divide>(Output<Node>&, Output<Node>&)>& Div) {
+    default_op_checks(node, 2, {"Div"});
+    auto x = node.get_input(0);
+    auto y = node.get_input(1);
+    auto result = Div(x, y);
+    set_node_name(node.get_name(), result);
+    return {result};
+}
+}  // namespace op
+}  // namespace tensorflow
+}  // namespace frontend
+}  // namespace ov
```
this file is no longer needed because you use `translate_binary_op` template that is sufficient

rkazants commented:
Hi @sami0i,

Sorry for the delay. I see that you have failures in the layer tests for `Div` operations you added. The problem is related to inference results OV and TF for this operation mismatch.
The reason of it is that you use `Divide` operation of OV that uses floor division by default.
Please implement separate function helper (as you did initially) `translate_div_op` and use `Divide` with `m_pythondiv` = false. See documentation: https://docs.openvino.ai/2023.2/openvino_docs_ops_arithmetic_Divide_1.html

Best regards,
Roman

rkazants reviewed (commented):
Please address comment I left

sami0i commented:
Hi @rkazants,
Thank you for your guidance.
I have a question: Should I delete the file `div.cpp` and place the `translate_div_op` function in the file `binary_op.cpp`?
Or, should I keep the file `div.cpp` and adapt the dictionary of Loaders based on that?

rkazants commented:
> Hi @rkazants, Thank you for your guidance. I have a question: Should I delete the file `div.cpp` and place the `translate_div_op` function in the file `binary_op.cpp`? Or, should I keep the file `div.cpp` and adapt the dictionary of Loaders based on that?

Let us use `div.cpp` and implement function helper there.

rkazants commented:
build_jenkins

rkazants commented:
Hi @sami0i,

Unfortunatelly, it still fails due to inference results mismatch between TF and OV:

```
E       AssertionError: Comparing with Framework failed: ie_res={'Div:0': array([[-1, -3, -1, -2, -1],
E              [-2,  3, -1, -3,  5],
E              [-1, -1,  1, -1, -1],
E              [ 6, -1, -2, -1, -5],
E              [-1, -1, -1,  0,  0],
E              [ 1, -2, -5,  0, -1],
E              [ 1,  0,  1, -2, -1],
E              [ 0,  6,  0, -1,  1]], dtype=int32), 'Div': array([[-1, -3, -1, -2, -1],
E              [-2,  3, -1, -3,  5],
E              [-1, -1,  1, -1, -1],
E              [ 6, -1, -2, -1, -5],
E              [-1, -1, -1,  0,  0],
E              [ 1, -2, -5,  0, -1],
E              [ 1,  0,  1, -2, -1],
E              [ 0,  6,  0, -1,  1]], dtype=int32)}; framework_res={'Div': array([[ 0, -3,  0, -1,  0],
E              [-1,  3,  0, -2,  5],
E              [ 0, -1,  1,  0,  0],
E              [ 6,  0, -2,  0, -5],
E              [-1,  0,  0,  0,  0],
E              [ 1, -2, -5,  0,  0],
E              [ 1,  0,  1, -1, -1],
E              [ 0,  6,  0,  0,  1]], dtype=int32)}.
```

Could you please investigate the logic for `Div`? And share examples when results diverge?
You can print generated operands in `_prepare_input` and share with me here. Now we understood that the single `Divide` does not fit well and more tricky sub-graph is needed for expression of `Div`.

Best regards,
Roman

sami0i commented:
Hi @rkazants,

I tested this many times, and unfortunately, I did not encounter any errors. For instance, the `Div` function was calculated correctly in the following example.

```
infer_res: 
{'Div:0': array([[ 1,  0,  0, -2,  1],
       [-1, -3, -2,  0,  2],
       [-1,  0, -2, -1, -1],
       [ 0,  0, -2, -1,  0],
       [ 0,  0,  2,  1, -1],
       [ 2,  0,  0,  0, -1],
       [ 1,  0,  0, -1,  0],
       [ 0,  0,  2,  1, -1]]), 'Div': array([[ 1,  0,  0, -2,  1],
       [-1, -3, -2,  0,  2],
       [-1,  0, -2, -1, -1],
       [ 0,  0, -2, -1,  0],
       [ 0,  0,  2,  1, -1],
       [ 2,  0,  0,  0, -1],
       [ 1,  0,  0, -1,  0],
       [ 0,  0,  2,  1, -1]])}
, framework_res: 
{'Div': array([[ 1,  0,  0, -2,  1],
       [-1, -3, -2,  0,  2],
       [-1,  0, -2, -1, -1],
       [ 0,  0, -2, -1,  0],
       [ 0,  0,  2,  1, -1],
       [ 2,  0,  0,  0, -1],
       [ 1,  0,  0, -1,  0],
       [ 0,  0,  2,  1, -1]])}
```

Could you please guide me on what I should do for this?

sami0i commented:
Hi @rkazants 

It appears that the error occurred only when calculating the `Div` on `int32` data types. I checked the `tf.raw_ops.Div` function, and it performs floor division for integer data types. Therefore, I think I should revert to setting `m_pythondiv` to `true`.

rkazants commented:
> Hi @rkazants
> 
> It appears that the error occurred only when calculating the `Div` on `int32` data types. I checked the `tf.raw_ops.Div` function, and it performs floor division for integer data types. Therefore, I think I should revert to setting `m_pythondiv` to `true`.

`m_pythondiv` to `true` is also not quite correct because we implement translators to support all expected types in order to be fully aligned with TF.

Best regards,
Roman

sami0i commented:
Hi @rkazants 

Based on this documentation: https://docs.openvino.ai/2023.2/openvino_docs_ops_arithmetic_Divide_1.html
The `m_pythondiv` attribute is supported only for integer data types. If set to true, `translate_to_div` calculates floor division for integer types and simple division for other types. This behavior is consistent with TF.

rkazants commented:
> Hi @rkazants
> 
> Based on this documentation: https://docs.openvino.ai/2023.2/openvino_docs_ops_arithmetic_Divide_1.html The `m_pythondiv` attribute is supported only for integer data types. If set to true, `translate_to_div` calculates floor division for integer types and simple division for other types. This behavior is consistent with TF.
> 
> Is it ok if I set `m_pythondiv` to `true` and check for errors in the test?

I am also unable to reproduce this inference mismatch. I tried the following and you can use this code for your local experiments to check some pair of numbers:

```
import numpy as np
import openvino.runtime.opset9 as ov
from openvino.runtime import Model, Core

# create a model with TopK
x = ov.parameter([1], name="x", dtype=np.int32)
y = ov.parameter([1], name="y", dtype=np.int32)
divide = ov.divide(x, y)
model = Model([divide], [x, y], "model")

# infer it on CPU
core = Core()
compiled_model = core.compile_model(model, "CPU")
ov_output = compiled_model.infer_new_request({0: np.array([4], dtype=np.int32), 1: np.array([-3], dtype=np.int32)})
print( "OV result = ", ov_output)

tf_output = tf.raw_ops.Div(x=np.array([4], dtype=np.int32),y=np.array([-3], dtype=np.int32))
print( "TF result = ", tf_output)
```

Let us return to `m_pythondiv` to true and see what happens. Also I recommend to print out values of operands to see it in CI.

Best regards,
Roman

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,49 @@
+        # generate x and y
+        inputs_data['x'] = np.random.randint(-10, 10, x_shape).astype(self.input_type)
+        inputs_data['y'] = np.random.randint(1, 10, y_shape)*np.random.choice([-1,1], y_shape)
+        return inputs_data
```
please print out `inputs_data['x']` and `inputs_data['y']` temporarily here.

sami0i commented:
Hi @rkazants 

Thank you for your guidance. I've modified the codes according to your comments.

rkazants commented:
build_jenkins

sami0i commented:
@rkazants 
Could it be that the new version of `div.cpp` wasn't built because of `clang`'s error? Is the unit test error related to the previous version of `div.cpp`?

rkazants commented:
Hi @sami0i,

Yes, you are right Windows is passing and Linux is failing:)

Here are inputs and outputs for failing case:

```
input x: 
[[-10  -7   9  -6  -5]
 [  8   0  -8   9   1]
 [ -5  -7  -6  -6   4]
 [ -1  -7  -8   5   7]
 [ -2   0  -4  -1   3]
 [ -3   7  -2   4  -6]
 [  2  -3  -1   3   2]
 [  1   2  -9   0  -7]]
input y: 
[[-6 -5 -2  3 -2]
 [-4 -4 -3  4 -4]
 [-1  3 -9 -7 -5]
 [-1  9 -8  1 -2]
 [-6  8 -4  6 -3]
 [-9  6  8 -3 -6]
 [ 8 -4  3 -3 -8]
 [ 8  9  3 -1  1]]

ie_res={'Div': array(
       [[ 1,  1, -5, -2,  2],
       [-2,  0,  2,  2, -1],
       [ 5, -3,  0,  0, -1],
       [ 1, -1,  1,  5, -4],
       [ 0,  0,  1, -1, -1],
       [ 0,  1, -1, -2,  1],
       [ 0,  0, -1, -1, -1],
       [ 0,  0, -3,  0, -7]], dtype=int32), 
framework_res={'Div': array(
      [[ 1,  1, -4, -2,  2],
       [-2,  0,  2,  2,  0],
       [ 5, -2,  0,  0,  0],
       [ 1,  0,  1,  5, -3],
       [ 0,  0,  1,  0, -1],
       [ 0,  1,  0, -1,  1],
       [ 0,  0,  0, -1,  0],
       [ 0,  0, -3,  0, -7]], dtype=int32)}.

```

rkazants commented:
Hi @sami0i,

It looks that TF `Div` divides operand and round the result to the nearest zero:

```
tf.raw_ops.Div(x=np.array([9], dtype=np.int32),y=np.array([-2], dtype=np.int32))
tf.raw_ops.Div(x=np.array([9], dtype=np.int32),y=np.array([2], dtype=np.int32))
```

So floor division is not suitable. Let us try to set `m_pythondiv` to false again and see cases where the test is failing. This way we understand the failing cases and handle them.

Best regards,
Roman

sami0i commented:
Hi @rkazants 

You're correct, `tf.raw_ops.Div` rounds the result to the nearest zero for integer data. I've set `m_pythondiv` to `false` again.

sami0i commented:
Hi @rkazants,

In a failing case, for example, with 9 divided by -2, the `ie_res` is -5, while the `framework-res` is -4. I tried these values using the code you mentioned earlier. There was no inference mismatch, and both results were -4. Additionally, `opset9.ov.divide` doesn't have a parameter for `m_pythondiv` that can be checked.

```
import numpy as np
import openvino.runtime.opset9 as ov
from openvino.runtime import Model, Core


data_type = np.int32
# create a model with TopK
x = ov.parameter([1], name="x", dtype=data_type)
y = ov.parameter([1], name="y", dtype=data_type)
divide = ov.divide(x, y)
model = Model([divide], [x, y], "model")

data_x = np.array([9], dtype=data_type)
data_y = np.array([-2], dtype=data_type)

# infer it on CPU
core = Core()
compiled_model = core.compile_model(model, "CPU")
ov_output = compiled_model.infer_new_request({0: data_x, 1: data_y})
print( "OV result = ", ov_output)

tf_output = tf.raw_ops.Div(x=data_x,y=data_y)
print( "TF result = ", tf_output)
```

rkazants commented:
build_jenkins

sami0i commented:
Hi @rkazants 

Unfortunately, the same error persists. Should I round the result to the nearest zero for integer data in `div.cpp`?

rkazants commented:
> Hi @rkazants
> 
> Unfortunately, the same error persists. Should I round the result to the nearest zero for integer data in `div.cpp`?

Hi @sami0i,

Sorry for the delay. I was thinking how to be. Let us extend your code for signed integer types that you can check like `x.get_element_type().is_integral_number() && x.get_element_type().is_signed()`. For this case, let us generate a boolen mask of elements for which `Mod` returns non-zero values and `Divide` result is negative, let us subtract the result by one for them. Please express this logic using graph operation representation.

Best regards,
Roman

sami0i commented:
Hi @rkazants 

Thank you for your guidance. I've updated div.cpp, but I didn't add `x.get_element_type().is_signed()` to the conditions because if either x or y is negative, the result of the division could also be negative.

rkazants commented:
build_jenkins

rkazants commented on src/frontends/tensorflow_common/src/op/div.cpp:
```diff
@@ -0,0 +1,55 @@
+    bool m_pythondiv = false;
+
+    // Check if the element type is a integer
+    if (x.get_element_type().is_integral_number()) {
```
please add additional check `x.get_element_type().is_signed()`. For unsigned, we should not have a problem

rkazants commented on src/frontends/tensorflow_common/src/op/div.cpp:
```diff
@@ -0,0 +1,55 @@
+    default_op_checks(node, 2, {"Div"});
+    auto x = node.get_input(0);
+    auto y = node.get_input(1);
+    bool m_pythondiv = false;
```
let us remove this attribute and use default `Div` as floor div

rkazants commented on src/frontends/tensorflow_common/src/op/div.cpp:
```diff
@@ -0,0 +1,55 @@
+    // Check if the element type is a integer
+    if (x.get_element_type().is_integral_number()) {
+        // prepare auxiliary zero constants of the same type as the inputs
+        auto zero = make_shared<v0::Constant>(element::i32, Shape{}, 0)->output(0);
+        zero = make_shared<v1::ConvertLike>(zero, x);
```
Please use this helper:

```
    auto const_zero = create_same_type_const_scalar<int32_t>(x, 0);
```

rkazants commented on src/frontends/tensorflow_common/src/op/div.cpp:
```diff
@@ -0,0 +1,55 @@
+        auto mask = make_shared<v1::LogicalAnd>(mod_non_zero, div_is_neg);
+
+        // add 1 to the divide result
+        auto one = make_shared<v0::Constant>(element::i32, Shape{}, 1)->output(0);
+        one = make_shared<v1::ConvertLike>(one, x);
```
please use `create_same_type_const_scalar`

rkazants reviewed (commented):
left a few comments. Other looks perfect.

sami0i commented:
Hi @rkazants,

I've modified `div.cpp` according to your comments. Should I remove the print `inputs_data['x']` and `inputs_data['y']` lines from the test file?

rkazants commented:
build_jenkins

rkazants commented:
> Hi @rkazants,
> 
> I've modified `div.cpp` according to your comments. Should I remove the print `inputs_data['x']` and `inputs_data['y']` lines from the test file?

Let us make sure that tests for `int` will pass and I will remove it. Hope to merge PR very soon.

Best regards,
Roman

sami0i commented:
Hi @rkazants 

I think It appears that  `<v1::Divide>` behaves inconsistently across various os. For instance, when dividing -8 by 3, we anticipated `<v1::Divide>` to yield -3. We then added 1 to the expected result to obtain -2. However, it seems that the function initially calculated -2 accurately, and by adding 1, we obtained -1 as a result.

The solution that I think could be correct is to initially convert integer and signed inputs to floats. After calculating the division, we can then return the floor for positive values and the ceiling for negative values. Finally, we can convert the result back to an integer type.

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,51 @@
+    test_data_basic = [
+        dict(input_shape=[10, 20], input_type=np.float32),
+        dict(input_shape=[2, 3, 4], input_type=np.float32),
+        dict(input_shape=[8, 5], input_type=np.int32),
```
```suggestion
        pytest.param(dict(input_shape=[8, 5], input_type=np.int32),
                              marks=pytest.mark.xfail(condition=platform.system() == 'Windows', reason='Ticket - TBD'))
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,52 @@
+        # generate x and y
+        inputs_data['x'] = np.random.randint(-10, 10, x_shape).astype(self.input_type)
+        inputs_data['y'] = np.random.randint(1, 10, y_shape)*np.random.choice([-1,1], y_shape)
+        print("input x: \n{}".format(inputs_data['x']))
+        print("input y: \n{}".format(inputs_data['y']))
```
```suggestion
```

rkazants commented:
> Hi @rkazants
> 
> I think It appears that `<v1::Divide>` behaves inconsistently across various os. For instance, when dividing -8 by 3, we anticipated `<v1::Divide>` to yield -3. We then added 1 to the expected result to obtain -2. However, it seems that the function initially calculated -2 accurately, and by adding 1, we obtained -1 as a result.
> 
> The solution that I think could be correct is to initially convert integer and signed inputs to floats. After calculating the division, we can then return the floor for positive values and the ceiling for negative values. Finally, we can convert the result back to an integer type.

Hi @sami0i,

It is a bug of OV that we have inconsistency for this op across different systems. Now I marked `int32` test case as expected failed for Windows. I will report this bug to my colleagues for further fixing.

Regarding your proposal to use floating-point type, we may have a problem for big integers that some precision can be lost during conversion to float.

Best regards,
Roman

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,50 @@
+    test_data_basic = [
+        dict(input_shape=[10, 20], input_type=np.float32),
+        dict(input_shape=[2, 3, 4], input_type=np.float32),
+        pytest.param(dict(input_shape=[8, 5], input_type=np.int32),
+                              marks=pytest.mark.xfail(condition=platform.system() == 'Windows', reason='Ticket - TBD'))
```
```suggestion
        pytest.param(dict(input_shape=[8, 5], input_type=np.int32),
                              marks=pytest.mark.xfail(condition=platform.system() == 'Windows', reason='Ticket - TBD')),
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,50 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import numpy as np
+import pytest
```
```suggestion
import platform
import pytest
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,51 @@
+        inputs_data = {}
+        # generate x and y
+        inputs_data['x'] = np.random.randint(-10, 10, x_shape).astype(self.input_type)
+        inputs_data['y'] = np.random.randint(1, 10, y_shape)*np.random.choice([-1,1], y_shape)
```
```suggestion
        inputs_data['y'] = np.random.randint(1, 10, y_shape)*np.random.choice([-1,1], y_shape)
        print("input x: \n{}".format(inputs_data['x']))
        print("input y: \n{}".format(inputs_data['y'])) 
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,53 @@
+        dict(input_shape=[10, 20], input_type=np.float32),
+        dict(input_shape=[2, 3, 4], input_type=np.float32),
+        pytest.param(dict(input_shape=[8, 5], input_type=np.int32),
+                              marks=pytest.mark.xfail(condition=platform.system() == 'Windows', reason='Ticket - TBD')),
```
```suggestion
                              marks=pytest.mark.xfail(reason='Ticket TBD - Divide inconsistent behavior on different systems')),
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import numpy as np
+import platform
```
```suggestion
```

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_Div.py:
```diff
@@ -0,0 +1,52 @@
+        # generate x and y
+        inputs_data['x'] = np.random.randint(-10, 10, x_shape).astype(self.input_type)
+        inputs_data['y'] = np.random.randint(1, 10, y_shape)*np.random.choice([-1,1], y_shape)
+        print("input x: \n{}".format(inputs_data['x']))
+        print("input y: \n{}".format(inputs_data['y'])) 
```
```suggestion
```

rkazants commented:
> Hi @rkazants
> 
> I think It appears that `<v1::Divide>` behaves inconsistently across various os. For instance, when dividing -8 by 3, we anticipated `<v1::Divide>` to yield -3. We then added 1 to the expected result to obtain -2. However, it seems that the function initially calculated -2 accurately, and by adding 1, we obtained -1 as a result.
> 
> The solution that I think could be correct is to initially convert integer and signed inputs to floats. After calculating the division, we can then return the floor for positive values and the ceiling for negative values. Finally, we can convert the result back to an integer type.

@sami0i, you are correct that Divide behaves inconsistently on different systems including both Windows and Linux. So I marked `int32` case as expected failed.

rkazants commented:
build_jenkins

rkazants commented:
build_jenkins
